src/shell/layout/mod.rs

```diff
@@ -126,10 +54,42 @@ pub fn is_dialog(window: &CosmicSurface) -> bool {
     false
 }
 
-pub fn has_floating_exception(window: &CosmicSurface) -> bool {
+#[derive(Debug, Clone, Default)]
+pub struct TilingExceptions {
+    app_ids: RegexSet,
+    titles: RegexSet,
+}
+
+impl TilingExceptions {
+    pub fn new(exceptions_config: &Vec<ApplicationException>) -> Self {
```

Comment:
`&Vec<_>` is a weird type for a function argument. It forces the caller to allocate specifically a `Vec`, while you only use it for iterating.

So I suggest something like this instead:
```suggestion
    pub fn new<'a, I>(exceptions_config: I) -> Self
    where
      I: Iterator<Item=&'a ApplicationException>
    {
```